torch/nn/modules/activation.py

```diff
@@ -687,12 +687,13 @@ def __init__(self, embed_dim, num_heads, dropout=0., bias=True, add_bias_kv=Fals
         self.head_dim = embed_dim // num_heads
         assert self.head_dim * num_heads == self.embed_dim, "embed_dim must be divisible by num_heads"
 
-        self.in_proj_weight = Parameter(torch.empty(3 * embed_dim, embed_dim))
-
         if self._qkv_same_embed_dim is False:
             self.q_proj_weight = Parameter(torch.Tensor(embed_dim, embed_dim))
             self.k_proj_weight = Parameter(torch.Tensor(embed_dim, self.kdim))
             self.v_proj_weight = Parameter(torch.Tensor(embed_dim, self.vdim))
+            self.register_parameter('in_proj_weight', None)
```

Comment:
Why register a non-existing parameter?

Reply:
I don't know how the DDP works. The in_proj_weight parameter was used in the forward func if query/key/value have same dimensions. Otherwise, the code switch to q/k/v_proj_weight.

Reply:
I think there is no need to do this TBH. There are 2 code paths in this module: either you use the same weight tensor, or you use 3 separate ones.

Reply:
Thanks. Let me try.

Reply:
So if I don't not register `in_proj_weight` as `None`, the unit test show the following error message.
```
ERROR:root:Caught exception: 'MultiheadAttention' object has no attribute 'in_proj_weight', exiting process with exit code: 10
```

Reply:
I think that's because the `forward` method always uses `in_proj_weight` regardless the value of  `_qkv_same_embed_dim `?

https://github.com/pytorch/pytorch/blob/fe08fbe20c84b24954bd7c6d396a1bfbc3c22ddf/torch/nn/modules/activation.py#L763-L779

Reply:
I will tell our users to use `find_unused_parameters` parameter. Thanks @pietern @mrshenli for teaching me DDP.

Reply:
That error seems fixable in the module itself as well. I think that's the real bug here, not the side effect that it causes unused parameters and has DDP complain about it.

Reply:
We have the `in_proj_weight` parameter there to avoid an BC breaking issue (the `forward` func in `nn.MultiheadAttention` module actually calls a multi_head_attention functional). We may be able to remove the unused param after we eventually fix the functional.

Reply:
I see -- what happens to the functional(s) if the parameter is `None` though?

Reply:
It's actually a problem of the functional's API design [here](https://github.com/pytorch/pytorch/blob/8ec041405391d999478d88ca936d0ba5bdf0ea71/torch/nn/functional.py#L3131). Ideally, if we put the `in_proj_weight` as a keyword argument (like `q_proj_weight`), we will have the flexibility to not create `in_proj_weight` in the `nn.MultiheadAttention` module.

So a quick fix, I thought, was to register `in_proj_weight` as `None` and pass it to the functional.

Reply:
If I read the code correctly, you could also hard-code a `None` for `in_proj_weights` in the call to the functional in the `forward` function.

Reply:
haha, that's very true. Thanks.

Then, if I want to move forward with this PR, where should I land the unit test? Or I could have a PR without unit test? Thanks.@pietern

Reply:
You could add a test (or two) that verify that the functional is called with `None` for `in_proj_weights` if the dims are not equal and called with `None` for the other tensors (and not `None` for `in_proj_weights`) if they are equal, perhaps?

This is definitely not a test that belongs in the distributed suite.